File: runCurvCluster.py
import numpy as np
import scipy as sc
import time as time
from tqdm import tqdm
from scipy.sparse import csr_matrix
from scipy.sparse.csgraph import connected_components as conncomp
from funs import ORcurvAll_sparse, distGeo, plotCluster, inputGraphs, varinfo

# parameters
T = np.logspace(1, 2, 20) # diffusion time scale
#T = np.linspace(1e-4, 2, 2)  
retEval = 0                    # num. retained evals (currently not relevant)
cutoff = 0.99              # fraction of mass to retain
lamb = 0              # regularising parameter (the larger the more accurate, but higher cost, and too large can blow up)
whichGraph = 3
precision = 1e-10
sample = 10
perturb = 0.05

# load graph ##WARNING currently only works for directed
(G,Aold,L,pos) = inputGraphs(whichGraph) 
     
# geodesic distances (this is OK for directed)
dist = distGeo(Aold)

# loop over all diffusion times
nComms = np.zeros([sample, len(T)]) 
vi = np.zeros(len(T))
for i in tqdm(range(len((T)))):
    #print(' Diffusion time: ', t)
    
    A = Aold.copy()

    # compute diffusion after time t[i]
    Phi_full = sc.sparse.linalg.expm(-T[i]*L.toarray())
    Phi = (np.max(Phi_full)*precision)*np.round(Phi_full / (np.max(Phi_full)*precision))

    # compute curvatures
    Kappa = ORcurvAll_sparse(G.edges,dist,Phi,cutoff,lamb)    
    
    # update edge curvatures
    for edge in G.edges:
        G.edges[edge]['kappa'] = Kappa[edge[0]][edge[1]]

    #cluster (remove edges with negative curv and find conn comps)    
    mink = np.min(Kappa)
    maxk = np.max(Kappa)

    labels = np.zeros([A.shape[0],sample])
    thres = np.random.normal(0, perturb*(maxk-mink), sample)

    for k in range(sample):
        ind = np.where(Kappa<=thres[k])     
        A[ind[0],ind[1]] = 0 #remove edges with -ve curv.       
        (nComms[k,i], labels[:,k]) = conncomp(csr_matrix(A, dtype=int), directed=False, return_labels=True)
    
    # compute VI
    (vi[i],_) = varinfo(labels);
    
    # plot
    plotCluster(G,T,pos,i,labels[:,0],vi[0:i+1],np.mean(nComms[:,0:i+1],axis=0))

File: runCurvCluster.m
addpath([pwd,'/functions']) % location of auxiliary functions

%% parameters
t = logspace(-2,3,20); % diffusion time scale 
prec = 0.99;           % %mass to retain (set to 1 for exact)
lambda = inf;          % entropy reg. parameter (set to 0 for exact)
movie = 1;             % create movie
sample = 20;
perturb = 0.1;

%% load graph
if ~exist('G','var') 
    [G,A,X,Y] = inputGraphs(13); %graph
end

f = figure('Visible',0,'Position',[100 100 1600 600]);

%% Compute geodesic distances
d = distGeo(sparse(A));
numcomms = zeros(length(t),sample); vi = zeros(1,length(t));
v = 0; 
for i = 1:length(t)
    disp(i)

    % compute diffusion after time t(i)
    [~, Phi] = distDiff(A,t(i));

    % compute curvatures
    [~,K] = ORcurvAll_sparse(A,d,Phi,prec,lambda);
        
    % update edge curvatures
    G.Edges.Kappa = K(tril(A)>0);
    
    % cluster  
    mink = min(G.Edges.Kappa); maxk = max(G.Edges.Kappa);
    comms = zeros(size(A,1),sample);
    for j = 1:sample
        r = randn*perturb*(maxk-mink);
        G1 = rmedge(G,find(G.Edges.Kappa <= r)); %remove edges with -ve curv.
        comms(:,j) = conncomp(G1)';
        numcomms(i,j) = max(comms(:,j));
    end
    
    % compute VI
    [vi(i),~] = varinfo(comms);
    
    if movie == 1
        % append frame to movie
        frame = plotcluster(G,t,mean(numcomms(1:i,:),2),comms,X,Y,vi(1:i),f); 
        if i < length(t); stopmov = 0; else; stopmov = 1; end
        v = createMovie(frame, v, stopmov); 
    end
end